cpp/src/pausable_thread_loop.cpp

```diff
@@ -50,47 +60,64 @@ PausableThreadLoop::~PausableThreadLoop() {
 }
 
 bool PausableThreadLoop::is_running() const noexcept {
-    std::lock_guard<std::mutex> lock(mutex_);
     return state_ != State::Paused;
 }
 
 void PausableThreadLoop::pause_nb() {
+    State expected = State::Running;
+    if (state_.compare_exchange_strong(
+            expected, State::Pausing, std::memory_order_acq_rel, std::memory_order_relaxed
+        ))
     {
-        std::lock_guard<std::mutex> lock(mutex_);
-        if (state_ == State::Running) {
-            state_ = State::Pausing;
-        }
+        state_.notify_all();
     }
-    cv_.notify_one();
 }
 
 void PausableThreadLoop::pause() {
     pause_nb();
-    // And wait for the loop to flip to paused state. Behaviour is
-    // undefined if someone else called resume/stop while we're in
-    // this function.
-    std::unique_lock lock(mutex_);
-    cv_.wait(lock, [this]() { return state_ == State::Paused; });
+    state_.wait(State::Pausing, std::memory_order_acquire);
```

Comment:
I don't understand the point of this function now. We flip the state to pausing and then wait for it to further flip to something else. Why is that useful?

Reply:
Yeah, I _think_ we want both. If the application doesn't care whether the pause has actually completed, then it can simply use `pause_nb` where appropriate.